Add Dockerfile for building binaries

moertel opened the pull request:
**Description:** This commit adds a Dockerfile that will build the OTEL binary.

**Link to tracking Issue:** _none_

**Testing:** Built successfully locally

**Documentation:** None, as it is a straightforward `docker build` and its behaviour is in-line with the convention of how a `Dockerfile` in a repo root is expected to work.

moertel commented:
This Dockerfile is what I've used to build an image from master for local use, and I figure it might be useful for others trying to build from source.

I wanted to put this out here, so we can discuss. Its functionality overlaps with that of https://github.com/open-telemetry/opentelemetry-collector/blob/master/cmd/otelcol/Dockerfile but I felt having two Dockerfiles where one only builds the binary and the other produces the Docker image is probably overkill.

However, it could be amended so only the binary is built and then put on the host. This would still enable others to build binaries easily but not interfere with `/cmd/otelcol/Dockerfile`. For ease of development, a `make` target could be introduced.

Thoughts?

owais commented:
We already have a docker file for building the final image inside ./cmd/otelcol. Can we update that file instead? We'll also need to update the Makefile to start building in docker going forward. I think it's a good step forward. 

My only concern is that repeated build inside docker will be extremely slow as all dependencies will be downloading every time. I'd suggest to add dependency caching to the first stage of the docker file and move it to `./cmd/otelcol/Dockerfile`.

owais commented:
If we do this, it'd still be worth it keeping make targets to allow building directly on host.

moertel commented:
@owais thanks for the feedback. It's in-line with the comment I posted (apparently a minute earlier), so I agree, there shouldn't be Dockerfiles with overlapping functionality. Personally, I expect to find a Dockerfile in the project root. Is there a reason to keep it in `/cmd/otelcol`?

Regarding the caching, I'm open to suggestions. I just pushed an update that downloads and caches the dependencies before it copies the repo's files. Is that what you had in mind? Or would you rather go with mounting a `.vendor` directory?

owais commented:
> Personally, I expect to find a Dockerfile in the project root. Is there a reason to keep it in /cmd/otelcol?

The project is structured in a way that it can produce multiple executables. Right now it produces only one and that is cmd/otelcol. In future there could be more under cmd/ and each one would have it's own docker file.

owais commented:
> Regarding the caching, I'm open to suggestions. I just pushed an update that downloads and caches the dependencies before it copies the repo's files. Is that what you had in mind? Or would you rather go with mounting a .vendor directory?

More like the later. Not vendor exactly but I'd mount a local directory as the GOPATH inside the docker container and tell git to ignore that directory. It wouldn't be as ideal as building natively but good enough IMO.

moertel commented:
Thanks for the additional input!

> My only concern is that repeated build inside docker will be extremely slow as all dependencies will be downloading every time.

We can't mount volumes during the `docker build` step (as the `$GOPATH` solution suggests) but I've addressed your concern by copying `go.mod` and `go.sum` over first. The `go mod download` layer will be cached unless there's a change in either file. I've also moved up the `make install-tools` step, so it too will be cached.


> The project is structured in a way that it can produce multiple executables. Right now it produces only one and that is cmd/otelcol. In future there could be more under cmd/ and each one would have it's own docker file.

Maybe we can have the best from both worlds. Here's another suggestion using build arguments in Docker. I've defaulted it to `target=otelcol`, so it would run `make otelcol`. As long as other executables follow the current naming scheme, this would work out of the box.

```
docker build --build-arg target=otelcol .
```

Which right now is equivalent to `docker build .` due to the default setting. A gotcha here is that I had to rename the executable to `otel` in order to make this work. We could theoretically pass on the build argument value to the entrypoint but at that point I found it too messy.

moertel commented:
Follow-up: I've found out about an [experimental `mount` feature](https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md#use-experimental-dockerfile-frontend) which does cache dependencies. However, as it's experimental at this point, we _probably_ shouldn't use it yet.

```diff
diff --git a/Dockerfile b/Dockerfile
index 77068f6..639137a 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -1,3 +1,4 @@
+# syntax=docker/dockerfile:experimental
 FROM alpine:3.10 as certs
 RUN apk --update add ca-certificates

@@ -14,7 +15,9 @@ RUN make install-tools

 ADD . .
 ARG target=otelcol
-RUN make ${target}
+RUN --mount=type=cache,target=/go/pkg/mod \
+    --mount=type=cache,target=/root/.cache/go-build \
+    make ${target}


 FROM scratch
```

owais commented:
> I've addressed your concern by copying go.mod and go.sum over first. The go mod download layer will be cached unless there's a change in either file.

Does this mean the cache will be invalidated every time either the go.sum or go.mod files change? I don't think that is desirable. Can we not just mount a local project directory into the build stage and set it as the GOPATH? If that doesn't work, may be it's not worth having a multi stage file and we should aim to have two files instead one for building collector and another for building the docker image.

moertel commented:
> Does this mean the cache will be invalidated every time either the go.sum or go.mod files change? I don't think that is desirable.

That's how the Docker cache is supposed to work, though.

> Can we not just mount a local project directory into the build stage and set it as the GOPATH?

See my comment https://github.com/open-telemetry/opentelemetry-collector/pull/538#issuecomment-584049899.

> [...] may be it's not worth having a multi stage file

It depends on the audience and here's where we disagree, I guess. For developers and contributors it's fine to have an actual dev environment set up on their computers and build locally; in that case I can see how frequent changes in `go.mod` become inconvenient in Docker in terms of build time. However, for someone visitors of your project (like me) who "just" want to get the `docker-compose` demo running or build an image from the latest `master`, it's a god-sent to have a `Dockerfile` that just does everything out of the box.

I'd like to understand the problem that this would cause on your end. Would you mind explaining why a slightly longer build time is an issue? I'd expect that you use `make otelcol` locally anyway, instead of Docker.

Your call, though. I'd advise against splitting up the process into two separate Dockerfiles.

moertel commented:
I'm still interested in merging this ;)

owais commented:
> That's how the Docker cache is supposed to work, though.

Sure, but if I understand correctly, it does not solve our use case.

It's not slightly longer at all. It takes considerably more time to build when dependencies change and this happens all the time during development. My point is that if we add docker based builds to the project, they should not cause a regression in the developer experience in anyway. Sure, we can assume that people who work on the project daily can continue to use native builds and occasional contributors can use docker build but I don't see the point of that. If we are providing dockerized builds, we should support that for everyone. Adding proper caching of dependencies using mounts does not degrade the experience for occasional developers in any way. 

Also, I don't think the Dockerfile should be the "interface" through which people build this project. It should always be the Makefile. Contributors and maintainers both should just reliable be able to run something like `make build` and have a good experience whether they do it 100 times a day or one a month.

moertel commented:
Fair enough, I give up then. The caching you're asking to be added doesn't exist, so there's nothing I can do, honestly. 

> [...] occasional contributors can use docker build but I don't see the point of that

FWIW, I wasted 2 hours of my life getting this project to build locally and I would have appreciated if there was a dockerised build that could have saved me this time. Not every (potential) contributor has the means or time to set up a functioning Go environment but right now there is no alternative. 🤷‍♀

owais commented:
I'm sorry for the bad experience you had. I still think we should add docker based builds. 

> Fair enough, I give up then. The caching you're asking to be added doesn't exist, so there's nothing I can do, honestly.

Probably I didn't communicate it clearly. All I was saying was that we should mount a local directory as a volume onto the build container and inside the build container set GOPATH to this mounted volume. That way GOPATH would be cached across builds and it would work for everyone. I think something like this should work:

```
docker run -v "$(PWD)":"/src" -v "$(PWD)/_caches/go":"/go" build-image "make otelcol"
```

If the build-image set `GOPATH` to `/go`, then `_caches/go` would be mounted there for every build and dependency caching would work always. I'm not sure if this would work with  multi-stage builds as is but this definitely works when using a separate build image.
